### PR TITLE
New model and admin interface for storing system wide prices

### PR DESCRIPTION
**Pull request by iant**
The BenchmarkMetrics class in the analytics includes some hard-coded prices:

```
  ELECTRICITY_PRICE = 0.15
  SOLAR_EXPORT_PRICE = 0.05
  GAS_PRICE = 0.03
  OIL_PRICE = 0.05
```

We want to move away from hard-coded prices and instead allow these to be managed by an EnergySparks admin via the application.  This PR
- [x] adds database fields to the existing settings model to manage a list of prices.
- [x] updates the existing settings admin interface to support changing values in the application UI
- [x] Add the following text as hints
Solar Price
used to calculate £ value of exporrted solar, with existing panels for the advice page
used to calculate economic benefit of installing solar for advice page and alert
Electricity Price
used to calculate cost of a kettle, smartphone, tv, console, for equivalences
used as default electricity tariff when school has gas only, when calculating cost of point of use water heaters
Gas price
use to calculate cost of a shower for equivalences

Supporting some level of caching (so as to avoid repeated database calls) and updating the analytics code to use saved values (or defaults as a fallback) will be handled in separate PR's.

### CODE REVIEW

File: app/models/site_settings.rb [hunk 1]
```diff
@@ -13,11 +13,29 @@
...
     order('created_at DESC').first || new
   end
 
   def temperature_recording_month_numbers
     temperature_recording_months.reject(&:blank?).map(&:to_i)
   end
+
+  def self.current_prices
+    Rails.cache.fetch(CURRENT_PRICES_CACHE_KEY) do
+      OpenStruct.new(current.prices)
```
Comment (iant) [inline app/models/site_settings.rb:32]: perhaps we need to default to constant values if store_accessor values are empty

**Comment by iant** [comment]
See also https://github.com/Energy-Sparks/energy-sparks_analytics/pull/356 - this would need to be be merged first.

File: app/models/site_settings.rb (hunk 1 above)
Comment (ldodds) [reply app/models/site_settings.rb:32]: I'd consider these values to be "seeds" so if we were starting from clean db then they'd need to be populated.

The migration adds defaults, so we should never be without the values.

File: app/models/site_settings.rb [hunk 2]
```diff
@@ -13,11 +13,29 @@
 #
 
 class SiteSettings < ApplicationRecord
+  store_accessor :prices, :electricity_price, :solar_export_price, :gas_price, :oil_price
+  validates :electricity_price, :solar_export_price, :gas_price, :oil_price, numericality: { only_float: true, allow_blank: true }
```
Comment (ldodds) [inline app/models/site_settings.rb:17]: We're allowing blank values here. Shouldn't we make sure there's always a price set?

**Review by ldodds — CHANGES_REQUESTED** [review]

File: app/models/site_settings.rb [hunk 3]
```diff
@@ -13,11 +13,29 @@
...
   def self.current
     order('created_at DESC').first || new
   end
 
   def temperature_recording_month_numbers
     temperature_recording_months.reject(&:blank?).map(&:to_i)
   end
+
+  def self.current_prices
+    Rails.cache.fetch(CURRENT_PRICES_CACHE_KEY) do
```
Comment (ldodds) [inline app/models/site_settings.rb:31]: What's the default cache time here? Should we be invalidating the cache when the SiteSettings are updated (if the prices have changed)

**Review by ldodds — CHANGES_REQUESTED** [review]

File: app/models/site_settings.rb (hunk 2 above)
Comment (iant) [reply app/models/site_settings.rb:17]: :+1:

File: app/models/site_settings.rb (hunk 3 above)
Comment (iant) [reply app/models/site_settings.rb:31]: no default cache time - we'll invalidate the cache after save (so runs both on create and update (just to be safe) - see `delete_current_prices_cache` l18)

File: app/models/site_settings.rb (hunk 2 above)
Comment (ldodds) [reply app/models/site_settings.rb:17]: So I think this is still outstanding. There must always be a non-zero price?

**Review by ldodds — CHANGES_REQUESTED** [review]

**Comment by iant** [comment]
@ldodds yep sorry - just pushed an update for non null validation and updated the test.

**Comment by iant** [comment]
^^ and fyi added a feature flag check to the analytics pr https://github.com/Energy-Sparks/energy-sparks_analytics/pull/356/files#diff-e4dd4e5bd5af7604cefb6b6617e0eb0eb93e0d56a7eefb3cf4b91de1cd441e53R24

**Comment by iant** [comment]
Putting a blocked label on this one for now while @PhilipTB takes a look at what needs doing in the analytics library to use other tariffs.

**Comment by iant** [comment]
@ldodds  is this ok to close now?  and ok close the related analytics pr https://github.com/Energy-Sparks/energy-sparks_analytics/pull/356 ?

File: db/migrate/20221103112116_add_prices_to_site_settings.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class AddPricesToSiteSettings < ActiveRecord::Migration[6.0]
+  def change
+    add_column(:site_settings, :prices, :jsonb)
+    SiteSettings.current.update!(
+      electricity_price: BenchmarkMetrics::ELECTRICITY_PRICE,
```
Comment (ldodds) [inline db/migrate/20221103112116_add_prices_to_site_settings.rb:5]: If this is merged before the analytics, then there won't be a price here?

Comment (iant) [reply db/migrate/20221103112116_add_prices_to_site_settings.rb:5]: indeed.  i'll hard code these values instead as they'll be over-writable anyway.  :+1:

**Comment by iant** [comment]
Screenshot 2023-05-09 at 14 16 46

File: app/views/schools/advice/solar_pv/_analysis_no_solar_table.html.erb [hunk 5]
```diff
@@ -50,6 +53,7 @@
         <td class="text-right"><%= format_unit(scenario.solar_pv_output_co2, :co2) %></td>
         <td class="text-right"><%= format_unit(scenario.capital_cost_£, :£) %></td>
         <td class="text-right"><%= scenario.payback_years.round %> years</td>
+        <td class="text-right"><%= format_unit(scenario.export_income_£, :£) %></td>
```
Comment (iant) [inline app/views/schools/advice/solar_pv/_analysis_no_solar_table.html.erb:56]: fyi @ldodds Export income is missing from this table (so no changes shown in the table with any changes in the site settings solar_export_price.  let me know if this needs moving/removing
Screenshot 2023-05-11 at 09 18 02

**Comment by iant** [comment]
@ldodds i think this ones ready for another review when you get a mo

**Review by ldodds — APPROVED** [review]

**Pull request merged**